Add Subsetting lb policy (#1)
s-matyukevich/proposal: A66-deterministic-subsetting-lb-policy.md

A66: `deterministic_subsetting` LB policy.
----
* Author(s): @s-matyukevich, @joybestourous
* Approver: 
* Status: Draft
* Implemented in: Go, Java
* Last updated: [Date]
* Discussion at:

## Abstract

Add support for the `deterministic_subsetting` load balancing policy configured via xDS.

## Background

> Currently, gRPC is lacking a way to select a subset of backend endpoints and load-balance requests between them. Mostly people are using two extremes: `pick_first` and `round_robin`. With `pick_first`, the resulting load distribution on the backend servers ends up being very uneven, especially during server rollouts. With `round_robin`, gRPC generates a lot of unnecessary connections, which consume resources on both clients and servers. The proposed LB policy provides a middle ground between these two extremes.

Reviewer: I think it'd be worth being a bit more precise, especially for why pick first is bad (it is counter intuitive to people that did not run into this problem):

```suggestion
Currently, gRPC is lacking a way to select a subset of endpoints available from the resolver and load-balance requests between them. Out of the box, users have the choice between two extremes: `pick_first` which sends all requests to one random backend, and `round_robin` which sends requests to all available backends. `pick_first` has poor connection balancing when the number of client is not much higher than the number of servers because of the birthday paradox. The problem is exacerbated during rollouts because `pick_first` does not change endpoint on resolver updates if the current subchannel remains `READY`. `round_robin` results in every servers having as many connections open as there are clients, which is unnecessarily costly when there are many clients, and makes local decisions load balancing (such as outlier detection) less precise.
```

### Related Proposals: 

## Proposal

Introduce a new LB policy, `deterministic_subsetting`. This policy selects a subset of addresses and passes them to the child LB policy in such a way that:

> * Every client is connected to at most N backends. If there are not enough backends, the policy falls back to its child policy.

Reviewer: It's "exactly" N backends if at least N backends are availbale, right?

```suggestion
* Every client is connected to exactly N backends. If there are not enough backends, the policy falls back to its child policy.
```

* The number of resulting connections per backend is as balanced as possible.

> * Every client is connected to a different subset of backends. 

Reviewer: I don't think you can guarantee that when there are few servers, but maybe you could rephrase in saying that it "has good subset diversity" (each client connects to a different subset of servers, as much as possible).

Author: Rephrased with `For any 2 given clients the probability of them connecting on the same or very similar subset of backends is as low as possible. `

```
This is the algorithm as previously described in Site Reliability Engineering: How Google Runs Production Systems, Chapter 20 but one improvement remains that can be made by balancing the leftover tasks in each group. The simplest way to achieve this is by choosing (before shuffling) which tasks will be leftovers in a round-robin fashion. For example, the first group of frontend tasks would choose {0, 1} to be leftovers and then shuffle the remaining tasks to get subsets {8, 3, 9, 2} and {4, 6, 5, 7}, and then the second group of frontend tasks would choose {2, 3} to be leftovers and shuffle the remaining tasks to get subsets {9, 7, 1, 6} and {0, 5, 4, 8}. This additional balancing ensures that all backend tasks are evenly excluded from consideration, producing a better distribution.
```

> 

Reviewer: I would insert pseudocode for the algorithm here. It will make it very easy for people to follow the implementation. Example of gRFC using pseudocode: https://github.com/grpc/proposal/blob/master/A60-xds-stateful-session-affinity-weighted-clusters.md#response-side-processing

Author: added some pseudocode with detailed comments.

### Characteristics of the selected algorithm

* Every client connects to exactly `subset_size` backends.

> * Clients are evenly distributed between backends. The most connected backend receives at most 2 more connections than the least connected one.

Reviewer: This is not clear why. I think you should explain it.

Author: added the explanation.

> A66: `deterministic_subsetting` LB policy.

Reviewer: See outstanding pull requests, I think it would be at least A68.

Reviewer: Ha, I think you fixed that but will want to delete this A66 file.

Author: yep, renamed it to A68

```

Here is the implementation of this algorithm in pseudocode with detailed comments. 

```
func filter_addresses(addresses, subset_size, client_index, sort_addresses)
    backend_count = addresses.length()
    if backend_count > subset_size {
        // if we don't have enough addresses to cover the desired subset size, just return the whole list
        return addresses
    }

    if sort_addresses {
        // sort address list because the algorithm assumes that the initial
        // order of the addresses is the same for every client
        addresses.sort()
    }

>     // subset_count indicates how much clients we can have so that every cleint is connected to exactly 

Reviewer: ```suggestion
    // subset_count indicates how many clients we can have so that every client is connected to exactly 
```

    // subset_size distinct backends and no 2 clients connect to the same backend.
    subset_count = backend_count / subset_size
    

>     // Given subset_count we not cat divide clients by rounds. Every round have exactly subset_count clients.

Reviewer: Not sure what you meant with the first sentence, but I guess you'll need to fix typos.

Author: `now can divide` fixing it now.

    // round indicates the index of the round for the current client based on its index.
    round = client_index / subset_count

>     // There might be some lefover backends withing every round in cases when backend_count % subset_size != 0

Reviewer: ```suggestion
    // There might be some lefover backends withing every round in cases when backend_count % subset_size != 0.
```

>     // excluded_count indicates how mumanych leftover backends we have on every round.

Reviewer: mumanych :)

Author: oops :) I should've run Ghat GPT again to correct all the typos. Thanks for looking at my comments.

    // Default is false.
    google.protobuf.BoolValue sort_addresses = 3;

    // The config for the child policy.
    repeated LoadBalancingConfig child_policy = 4;
}
```

### Handling Parent/Resolver Updates

When the resolver updates the list of addresses, or the LB config changes, Deterministic subsetting LB will run the subsetting algorithm, described above, to filter the endpoint list. Then it will create a new resolver state with the filtered list of the addresses and pass it to the child LB. Attributes and service config from the old resolver state will be copied to the new one. 

### Handling Subchannel Connectivity State Notifications

Deterministic subsetting LB will simply redirect all requests to the child LB without doing any additional processing. This also applies to all other callbacks in the LB interface, besides the one that handles resolver and config updates (which is described in the previous section). This is possible because deterministic subsetting LB doesn't store or manage sub-connections - it acts as a simple filter on the resolver state, and that's why it can redirect all actual work to the child LB. 

### xDS Integration

> Deterministic subsetting LB won't depend on xDS in any way. People may choose to initialize it by directly providing service config. If they do this, they will have to figure out how to correctly initialize and keep updating client\_index. There are ways to do that, but describing them is out of scope for this gRPC. The main point is that the LB itself will be xDS agnostic.

Reviewer 2: ```suggestion
Deterministic subsetting LB won't depend on xDS in any way. People may choose to initialize it by directly providing service config. If they do this, they will have to figure out how to correctly initialize and keep updating client\_index. There are ways to do that, but describing them is out of scope for this gRFC. The main point is that the LB itself will be xDS agnostic.
```

Think you meant gRFC in this context!
